File: loopx/control_plane/goals/shared_goal_alignment.py
"""Read-only shared goal alignment projection adapter (RFC Stage 1).

This adapter collects typed facts for one registered Agent around one shared
Goal — registry identity, the selected Todo/lease source, the append-only state
event log, Todo claim/lease fields, and recorded replan obligations — and
asks the TypeScript-owned reducer (``goal.shared_goal_alignment.project``)
to project ``shared_goal_alignment_v0``.

Derivation invariants (RFC shared-goal-alignment-and-governed-amendment-v0
§3.3): every projected field is derived from typed facts only. Shared
``Next Action`` prose, agent vision prose, and chat prose are never inputs.

The projection is strictly read-only: no writer path is touched, and no
approval or escalation semantics exist here. ``source_basis_digest`` is a
typed source-facts basis summary (goal status, registered agents, and
event-log basis facts, and canonical Todo revision when promoted), not a canonical intent-envelope digest — the full
RFC §3.1 envelope (objective, non-goals, acceptance, permission scope,
terminal conditions) has no typed storage yet, so nothing here claims
canonical intent identity.

Basis semantics: the only goal-level monotonic sequence carrier on this
codebase is the state event log's ``append_sequence``, so
``state_event_basis_sequence`` reports that event projection basis — it is
NOT a canonical goal/intent revision. Goals without a parsable
``events.jsonl`` use sequence 0 and an unbound Agent frontier. Before promotion
this is ``markdown_active_state``; after promotion it is ``canonical_todo_snapshot``
with a separate ``todo_basis`` token, never a fabricated event sequence;
drift is then reported as ``frontier_basis_unverifiable`` instead of a
fabricated behind fact.
"""

from __future__ import annotations

import hashlib
import json
from collections.abc import Mapping
from pathlib import Path
from typing import Any

from ...agent_registry import registered_agent_ids_for_goal
from ...event_sourced_state import (
    AppendOnlyStateEventStore,
    StateEventError,
    build_state_projection,
    event_sort_key,
)
from ...registry import registry_goals
from ..effect_runtime import EffectRuntimeRejected, effect_runtime_result
from ..todos.contract import normalize_todo_claimed_by
from .shared_goal_work_source import SharedGoalWorkSource, read_shared_goal_work_source
from .active_state_event_projection import state_event_log_candidates
from .active_state_metadata import parse_state_frontmatter
from .goal_frontier import (
    autonomous_replan_is_required,
    select_autonomous_replan_obligation,
)
from .path_resolution import resolve_goal_local_path

SHARED_GOAL_ALIGNMENT_EFFECT_METHOD = "goal.shared_goal_alignment.project"
SHARED_GOAL_ALIGNMENT_REQUEST_SCHEMA_VERSION = "shared_goal_alignment_request_v0"
SHARED_GOAL_ALIGNMENT_SCHEMA_VERSION = "shared_goal_alignment_v0"
REVISION_BASIS_STATE_EVENT_LOG = "state_event_log"
REVISION_BASIS_MARKDOWN_ACTIVE_STATE = "markdown_active_state"
BASIS_SOURCE_STATE_EVENT_LOG = "state_event_log"
BASIS_SOURCE_UNBOUND = "unbound"
DEFAULT_REGISTRY_RELATIVE_PATH = Path(".loopx") / "registry.json"


def _canonical_digest(value: object) -> str:
    # Local copy of the repository digest recipe: control_plane must not
    # import loopx.capabilities (m6 control_plane_outward_dependency).
    encoded = json.dumps(
        value, ensure_ascii=False, sort_keys=True, separators=(",", ":")
    ).encode("utf-8")
    return "sha256:" + hashlib.sha256(encoded).hexdigest()


def _registered_goal(
    registry_payload: Mapping[str, Any],
    *,
    goal_id: str,
) -> dict[str, Any]:
    for goal in registry_goals(dict(registry_payload)):
        if str(goal.get("id") or "") == goal_id:
            return goal
    raise ValueError(f"goal is not registered: {goal_id}")


def _load_state_event_facts(
    goal: Mapping[str, Any],
    *,
    state_path: Path,
) -> dict[str, Any] | None:
    """Load the first parsable state event log for the goal, read-only."""

    for event_log_path in state_event_log_candidates(
        dict(goal),
        state_path=state_path,
        resolve_goal_local_path=resolve_goal_local_path,
    ):
        if not event_log_path.exists():
            continue
        try:
            events = AppendOnlyStateEventStore(event_log_path).load()
            if not events:
                continue
            projection = build_state_projection(
                events,
                goal_id=str(goal.get("id") or "") or None,
            )
        except (OSError, StateEventError):
            continue
        return {"events": events, "projection": projection}
    return None


def _agent_frontier_basis(
    event_facts: Mapping[str, Any] | None,
    *,
    agent_id: str,
) -> dict[str, Any]:
    """Derive the Agent's frontier basis from its own attributed events.

    ``based_on_state_event_sequence`` is the highest append sequence among
    events whose ``actor_agent_id`` belongs to this Agent. Events attributed
    to peers never advance another Agent's basis.
    """

    events = event_facts.get("events") if event_facts else None
    if not isinstance(events, list):
        return {
            "based_on_state_event_sequence": None,
            "basis_source": BASIS_SOURCE_UNBOUND,
            "last_agent_event_id": None,
        }
    based_on: int | None = None
    last_agent_event_id: str | None = None
    for event in sorted(
        (item for item in events if isinstance(item, dict)),
        key=event_sort_key,
    ):
        actor = normalize_todo_claimed_by(event.get("actor_agent_id"))
        if actor != agent_id:
            continue
        try:
            sequence = int(event.get("append_sequence") or 0)
        except (TypeError, ValueError):
            continue
        if sequence < 1:
            continue
        based_on = sequence
        event_id = str(event.get("event_id") or "").strip()
        last_agent_event_id = event_id or None
    if based_on is None:
        return {
            "based_on_state_event_sequence": None,
            "basis_source": BASIS_SOURCE_UNBOUND,
            "last_agent_event_id": None,
        }
    return {
        "based_on_state_event_sequence": based_on,
        "basis_source": BASIS_SOURCE_STATE_EVENT_LOG,
        "last_agent_event_id": last_agent_event_id,
    }


def _source_basis_facts_envelope(
    *,
    goal_id: str,
    goal_status: str | None,
    registered_agents: list[str],
    revision_basis: str,
    last_append_sequence: int | None,
    source_checksum: str | None,
    state_updated_at: str | None,
    todo_basis: dict[str, Any] | None = None,
) -> dict[str, Any]:
    return {
        "goal_id": goal_id,
        "goal_status": goal_status,
        "registered_agents": sorted(registered_agents),
        "revision_basis": revision_basis,
        "last_append_sequence": last_append_sequence,
        "source_checksum": source_checksum,
        "state_updated_at": state_updated_at,
        **({"todo_basis": todo_basis} if todo_basis is not None else {}),
    }


def project_shared_goal_alignment(
    *, goal_id: str, agent_id: str | None, project: Path,
    registry_path: Path | None = None, runtime_root: Path | None = None,
    status_item: Mapping[str, Any] | None = None,
    project_asset: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    """Public read-only entrypoint; callers cannot inject the Todo snapshot."""
    return _project_shared_goal_alignment(goal_id=goal_id, agent_id=agent_id,
        project=project, registry_path=registry_path, runtime_root=runtime_root,
        status_item=status_item, project_asset=project_asset)


def _project_shared_goal_alignment(
    *,
    goal_id: str,
    agent_id: str | None,
    project: Path,
    registry_path: Path | None = None,
    runtime_root: Path | None = None,
    status_item: Mapping[str, Any] | None = None,
    project_asset: Mapping[str, Any] | None = None,
    work_source: SharedGoalWorkSource | None = None,
) -> dict[str, Any]:
    """Project the read-only ``shared_goal_alignment_v0`` view for one Agent."""

    normalized_goal_id = str(goal_id or "").strip()
    if not normalized_goal_id:
        raise ValueError("goal_id must be a non-empty registered goal id")
    normalized_agent_id = normalize_todo_claimed_by(agent_id)
    if not normalized_agent_id:
        raise ValueError("agent_id must be a public-safe agent id")

    effective_registry_path = (
        registry_path if registry_path is not None
        else project / DEFAULT_REGISTRY_RELATIVE_PATH
    )
    try:
        registry_payload = json.loads(
            effective_registry_path.read_text(encoding="utf-8")
        )
    except (OSError, ValueError):
        raise ValueError(
            f"goal registry is unreadable: {effective_registry_path}"
        ) from None
    if not isinstance(registry_payload, dict):
        raise ValueError("goal registry must contain a JSON object")
    goal = _registered_goal(registry_payload, goal_id=normalized_goal_id)

    registered_agents = registered_agent_ids_for_goal(goal)
    if normalized_agent_id not in registered_agents:
        raise ValueError(
            f"agent is not registered for goal {normalized_goal_id}: "
            f"{normalized_agent_id}"
        )

    effective_runtime_root = runtime_root if runtime_root is not None else _runtime_root_from_registry(registry_payload)
    source = work_source or read_shared_goal_work_source(
        goal=goal, project=project, runtime_root=effective_runtime_root,
    )
    if source.goal_id != normalized_goal_id:
        raise ValueError("shared work snapshot belongs to another Goal")
    state_file, state_text = source.state_path, source.state_text

    event_facts = _load_state_event_facts(goal, state_path=state_file)
    frontmatter = parse_state_frontmatter(state_text)
    state_updated_at = str(frontmatter.get("updated_at") or "").strip() or None
    goal_status = str(goal.get("status") or "").strip() or (
        str(frontmatter.get("status") or "").strip() or None
    )

    if event_facts is not None:
        projection = event_facts["projection"]
        revision_basis = REVISION_BASIS_STATE_EVENT_LOG
        try:
            basis_sequence = int(projection.get("last_append_sequence") or 0)
        except (TypeError, ValueError):
            basis_sequence = 0
        source_checksum = (
            str(projection.get("source_checksum") or "").strip() or None
        )
    else:
        revision_basis = ("canonical_todo_snapshot" if source.canonical_basis is not None
            else REVISION_BASIS_MARKDOWN_ACTIVE_STATE)
        basis_sequence = 0
        source_checksum = None

    source_basis_digest = _canonical_digest(
        _source_basis_facts_envelope(
            goal_id=normalized_goal_id,
            goal_status=goal_status,
            registered_agents=registered_agents,
            revision_basis=revision_basis,
            last_append_sequence=(
                basis_sequence
                if revision_basis == REVISION_BASIS_STATE_EVENT_LOG
                else None
            ),
            source_checksum=source_checksum,
            state_updated_at=state_updated_at,
            todo_basis=source.canonical_basis,
        )
    )
    source_basis = {
        "state_event_basis_sequence": basis_sequence,
        "source_basis_digest": source_basis_digest,
        "revision_basis": revision_basis,
        "state_updated_at": state_updated_at,
        **({"todo_basis": source.canonical_basis} if source.canonical_basis is not None else {}),
    }

    frontier_basis = _agent_frontier_basis(
        event_facts,
        agent_id=normalized_agent_id,
    )

    replan_obligation = select_autonomous_replan_obligation(
        dict(status_item) if isinstance(status_item, Mapping) else {},
        dict(project_asset) if isinstance(project_asset, Mapping) else None,
        agent_id=normalized_agent_id,
    )

    request = {
        "schema_version": SHARED_GOAL_ALIGNMENT_REQUEST_SCHEMA_VERSION,
        "goal_id": normalized_goal_id,
        "agent_id": normalized_agent_id,
        "source_basis": source_basis,
        "frontier_basis": frontier_basis,
        "work_items": source.items,
        "observed_at": source.observed_at,
        "open_lane_replan_obligation_required": (
            autonomous_replan_is_required(replan_obligation)
        ),
    }
    try:
        result = effect_runtime_result(
            SHARED_GOAL_ALIGNMENT_EFFECT_METHOD,
            request,
        )
    except EffectRuntimeRejected as exc:
        raise ValueError(str(exc)) from None
    if not isinstance(result, Mapping) or (
        result.get("schema_version") != SHARED_GOAL_ALIGNMENT_SCHEMA_VERSION
        or result.get("read_only") is not True
        or result.get("goal_id") != normalized_goal_id
        or result.get("agent_id") != normalized_agent_id
        or not isinstance(result.get("drift_facts"), list)
        or not isinstance(result.get("conflict_facts"), list)
    ):
        raise RuntimeError("TypeScript shared goal alignment shape mismatch")
    return dict(result)


def _runtime_root_from_registry(
    registry_payload: Mapping[str, Any],
) -> Path | None:
    raw = registry_payload.get("common_runtime_root")
    text = str(raw or "").strip()
    return Path(text).expanduser() if text else None

File: loopx/control_plane/goals/shared_goal_alignment.ts
import type { JsonObject } from "../effect_program.ts";
import { sharedGoalWorkFacts } from "./shared_goal_work.ts";
import { EffectRuntimeRequestError } from "../effect_runtime_errors.ts";
import {
  optionalNonEmptyString,
  requireBoolean,
  requireInteger,
  requireJsonObject,
  requireNonEmptyString,
  requireStringArray,
  requireStringLiteral,
} from "../runtime_decode.ts";

/**
 * Read-only `shared_goal_alignment_v0` projection contract (RFC
 * shared-goal-alignment-and-governed-amendment-v0, Stage 1).
 *
 * The reducer is deterministic: every projected field is derived from the
 * typed facts in the request. Shared `Next Action` prose, agent vision, and
 * chat prose are never inputs. The source basis sequence, digest, and
 * frontier basis are decoded and passed through unchanged — the digest is
 * computed once on the Python side so both runtimes observe one value.
 *
 * Basis semantics: `source_basis` is an event-log-derived projection basis,
 * NOT a canonical intent revision. With no event log, canonical Todo reads
 * use `canonical_todo_snapshot`, sequence 0 and an unbound Agent frontier.
 * The separate `todo_basis` identifies the single Todo/lease snapshot. `state_event_basis_sequence` is the state
 * event log's append sequence (or 0 with the markdown fallback), and
 * `source_basis_digest` hashes goal status, registered agents, event-log
 * basis facts and the canonical Todo revision when promoted — the RFC §3.1 canonical intent envelope (objective,
 * non-goals, acceptance, permissions, terminal conditions) has no typed
 * storage yet, so no field here claims canonical intent identity. This
 * contract has no writer surface: it projects drift and conflict facts only
 * and always answers with `read_only: true`.
 */

export const SHARED_GOAL_ALIGNMENT_REQUEST_SCHEMA_VERSION =
  "shared_goal_alignment_request_v0";
export const SHARED_GOAL_ALIGNMENT_SCHEMA_VERSION = "shared_goal_alignment_v0";

export const SHARED_GOAL_ALIGNMENT_DRIFT_FACTS = [
  "frontier_basis_behind",
] as const;
export const SHARED_GOAL_ALIGNMENT_CONFLICT_FACTS = [
  "frontier_basis_unverifiable",
  "lease_owner_mismatch",
  "open_lane_replan_obligation",
  "peer_claimed_lane_conflict",
] as const;

export type SharedGoalAlignmentDriftFact =
  (typeof SHARED_GOAL_ALIGNMENT_DRIFT_FACTS)[number];
export type SharedGoalAlignmentConflictFact =
  (typeof SHARED_GOAL_ALIGNMENT_CONFLICT_FACTS)[number];

const REVISION_BASIS_VALUES = [
  "state_event_log",
  "markdown_active_state",
  "canonical_todo_snapshot",
] as const;
const BASIS_SOURCE_VALUES = ["state_event_log", "unbound"] as const;

const TODO_ID_PATTERN = /^todo_[a-z0-9_-]{3,64}$/;
const AGENT_ID_PATTERN = /^[a-z][a-z0-9_.:@-]{0,79}$/;
// Mirrors loopx.runtime.validate_goal_id_path_segment (and the loopx.history
// copy): a goal id is any non-empty, safe single filesystem path segment —
// not "." or "..", and never containing "/", "\", or whitespace. The
// repository Goal-ID contract does not require a "goal-" prefix; registered
// goal ids such as "loopx-meta" must decode.
const GOAL_ID_PATTERN = /^(?!\.\.?$)[^\s/\\]+$/;
const SOURCE_BASIS_DIGEST_PATTERN = /^sha256:[0-9a-f]{64}$/;

export type RevisionBasis = (typeof REVISION_BASIS_VALUES)[number];
export type BasisSource = (typeof BASIS_SOURCE_VALUES)[number];

export interface SourceBasisFacts extends JsonObject {
  state_event_basis_sequence: number;
  source_basis_digest: string;
  revision_basis: RevisionBasis;
  state_updated_at: string | null;
  todo_basis?: JsonObject;
}

export interface FrontierBasisFacts extends JsonObject {
  based_on_state_event_sequence: number | null;
  basis_source: BasisSource;
  last_agent_event_id: string | null;
}

export interface FrontierCounts extends JsonObject {
  current_agent_claimed_advancement_count: number;
  unclaimed_advancement_count: number;
  other_agent_claimed_advancement_count: number;
}

export interface AlignmentClaimFacts extends JsonObject {
  todo_id: string;
  claimed_by: string;
  lease_epoch: number | null;
  lease_owner: string | null;
}

export interface UnclaimedEligibleFacts extends JsonObject {
  todo_id: string;
  task_class: "advancement_task";
  action_kind?: string;
}

export interface SharedGoalAlignmentRequest extends JsonObject {
  schema_version: typeof SHARED_GOAL_ALIGNMENT_REQUEST_SCHEMA_VERSION;
  goal_id: string;
  agent_id: string;
  source_basis: SourceBasisFacts;
  frontier_basis: FrontierBasisFacts;
  frontier_counts: FrontierCounts;
  claims: AlignmentClaimFacts[];
  unclaimed_eligible: UnclaimedEligibleFacts[];
  peer_claimed_bound_todo_ids: string[];
  open_lane_replan_obligation_required: boolean;
}

export interface SharedGoalAlignment extends JsonObject {
  schema_version: typeof SHARED_GOAL_ALIGNMENT_SCHEMA_VERSION;
  goal_id: string;
  agent_id: string;
  source_basis: SourceBasisFacts;
  frontier_basis: FrontierBasisFacts;
  frontier_counts: FrontierCounts;
  unclaimed_eligible_work: Array<{
    todo_id: string;
    claim_required_before_work: true;
  }>;
  drift_facts: SharedGoalAlignmentDriftFact[];
  conflict_facts: SharedGoalAlignmentConflictFact[];
  read_only: true;
}

function agentId(value: unknown, label: string): string {
  const decoded = requireNonEmptyString(value, label).trim().toLowerCase();
  if (!AGENT_ID_PATTERN.test(decoded)) {
    throw new EffectRuntimeRequestError(
      `${label} must be a public-safe agent id`,
    );
  }
  return decoded;
}

function todoId(value: unknown, label: string): string {
  const decoded = requireNonEmptyString(value, label).trim().toLowerCase();
  if (!TODO_ID_PATTERN.test(decoded)) {
    throw new EffectRuntimeRequestError(`${label} must be a valid Todo id`);
  }
  return decoded;
}

function nonNegativeInteger(value: unknown, label: string): number {
  const decoded = requireInteger(value, label);
  if (decoded < 0) {
    throw new EffectRuntimeRequestError(`${label} must be non-negative`);
  }
  return decoded;
}

function requireNoDuplicates(values: string[], label: string): void {
  const seen = new Set<string>();
  for (const value of values) {
    if (seen.has(value)) {
      throw new EffectRuntimeRequestError(
        `${label} contains a duplicate todo_id: ${value}`,
      );
    }
    seen.add(value);
  }
}

function decodeSourceBasis(value: unknown): SourceBasisFacts {
  const raw = requireJsonObject(value, "shared_goal_alignment.source_basis");
  const revisionBasis = requireStringLiteral(
    raw.revision_basis,
    REVISION_BASIS_VALUES,
    "shared_goal_alignment.source_basis.revision_basis",
    "shared_goal_alignment source_basis revision_basis is unsupported",
  );
  const basisSequence = nonNegativeInteger(
    raw.state_event_basis_sequence,
    "shared_goal_alignment.source_basis.state_event_basis_sequence",
  );
  if (revisionBasis === "state_event_log" && basisSequence < 1) {
    throw new EffectRuntimeRequestError(
      "shared_goal_alignment source_basis state_event_basis_sequence must be a positive event append sequence when revision_basis is state_event_log",
    );
  }
  if (revisionBasis !== "state_event_log" && basisSequence !== 0) {
    throw new EffectRuntimeRequestError(
      `shared_goal_alignment source_basis state_event_basis_sequence must be 0 when revision_basis is ${revisionBasis}`,
    );
  }
  const sourceBasisDigest = requireNonEmptyString(
    raw.source_basis_digest,
    "shared_goal_alignment.source_basis.source_basis_digest",
  );
  if (!SOURCE_BASIS_DIGEST_PATTERN.test(sourceBasisDigest)) {
    throw new EffectRuntimeRequestError(
      "shared_goal_alignment.source_basis.source_basis_digest must be a sha256:<hex> digest computed from typed source facts",
    );
  }
  let todoBasis: JsonObject | undefined;
  if (raw.todo_basis !== undefined) {
    const basis = requireJsonObject(raw.todo_basis, "todo_basis");
    if (basis.source_authority !== "file_v0" ||
      typeof basis.records_sha256 !== "string" || !/^[a-f0-9]{64}$/.test(basis.records_sha256)) {
      throw new EffectRuntimeRequestError("invalid canonical Todo basis");
    }
    todoBasis = {source_authority: basis.source_authority, records_sha256: basis.records_sha256,
      provider_revision: requireNonEmptyString(basis.provider_revision, "todo_basis.provider_revision")};
  }
  if (revisionBasis === "canonical_todo_snapshot" && todoBasis === undefined) {
    throw new EffectRuntimeRequestError("canonical_todo_snapshot requires a canonical Todo basis");
  }
  return {
    state_event_basis_sequence: basisSequence,
    source_basis_digest: sourceBasisDigest,
    revision_basis: revisionBasis,
    state_updated_at: optionalNonEmptyString(
      raw.state_updated_at,
      "shared_goal_alignment.source_basis.state_updated_at",
    ),
    ...(todoBasis === undefined ? {} : {todo_basis: todoBasis}),
  };
}

function decodeFrontierBasis(
  value: unknown,
  sourceBasis: SourceBasisFacts,
): FrontierBasisFacts {
  const raw = requireJsonObject(value, "shared_goal_alignment.frontier_basis");
  const basisSource = requireStringLiteral(
    raw.basis_source,
    BASIS_SOURCE_VALUES,
    "shared_goal_alignment.frontier_basis.basis_source",
    "shared_goal_alignment frontier basis_source is unsupported",
  );
  if (basisSource === "state_event_log") {
    if (sourceBasis.revision_basis !== "state_event_log") {
      throw new EffectRuntimeRequestError(
        "shared_goal_alignment frontier basis_source state_event_log cannot be compared against a markdown_active_state source basis",
      );
    }
    const basedOn = requireInteger(
      raw.based_on_state_event_sequence,
      "shared_goal_alignment.frontier_basis.based_on_state_event_sequence",
    );
    if (basedOn < 1) {
      throw new EffectRuntimeRequestError(
        "shared_goal_alignment.frontier_basis.based_on_state_event_sequence must be a positive event append sequence when basis_source is state_event_log",
      );
    }
    return {
      based_on_state_event_sequence: basedOn,
      basis_source: basisSource,
      last_agent_event_id: optionalNonEmptyString(
        raw.last_agent_event_id,
        "shared_goal_alignment.frontier_basis.last_agent_event_id",
      ),
    };
  }
  if (
    raw.based_on_state_event_sequence !== null &&
    raw.based_on_state_event_sequence !== undefined
  ) {
    throw new EffectRuntimeRequestError(
      "shared_goal_alignment.frontier_basis.based_on_state_event_sequence must be null when basis_source is unbound",
    );
  }
  if (
    raw.last_agent_event_id !== null &&
    raw.last_agent_event_id !== undefined
  ) {
    throw new EffectRuntimeRequestError(
      "shared_goal_alignment.frontier_basis.last_agent_event_id must be null when basis_source is unbound",
    );
  }
  return {
    based_on_state_event_sequence: null,
    basis_source: basisSource,
    last_agent_event_id: null,
  };
}

function decodeFrontierCounts(value: unknown): FrontierCounts {
  const raw = requireJsonObject(value, "shared_goal_alignment.frontier_counts");
  return {
    current_agent_claimed_advancement_count: nonNegativeInteger(
      raw.current_agent_claimed_advancement_count,
      "shared_goal_alignment.frontier_counts.current_agent_claimed_advancement_count",
    ),
    unclaimed_advancement_count: nonNegativeInteger(
      raw.unclaimed_advancement_count,
      "shared_goal_alignment.frontier_counts.unclaimed_advancement_count",
    ),
    other_agent_claimed_advancement_count: nonNegativeInteger(
      raw.other_agent_claimed_advancement_count,
      "shared_goal_alignment.frontier_counts.other_agent_claimed_advancement_count",
    ),
  };
}

function decodeClaims(
  value: unknown,
  agentIdValue: string,
): AlignmentClaimFacts[] {
  if (!Array.isArray(value)) {
    throw new EffectRuntimeRequestError(
      "shared_goal_alignment.claims must be an array",
    );
  }
  const claims: AlignmentClaimFacts[] = value.map((item, index) => {
    const raw = requireJsonObject(
      item,
      `shared_goal_alignment.claims[${index}]`,
    );
    const claimedBy = agentId(
      raw.claimed_by,
      `shared_goal_alignment.claims[${index}].claimed_by`,
    );
    if (claimedBy !== agentIdValue) {
      throw new EffectRuntimeRequestError(
        `shared_goal_alignment.claims[${index}].claimed_by must match the projected agent's own claim`,
      );
    }
    const leaseEpoch =
      raw.lease_epoch === null || raw.lease_epoch === undefined
        ? null
        : requireInteger(
            raw.lease_epoch,
            `shared_goal_alignment.claims[${index}].lease_epoch`,
          );
    if (leaseEpoch !== null && leaseEpoch < 1) {
      throw new EffectRuntimeRequestError(
        `shared_goal_alignment.claims[${index}].lease_epoch must be a positive lease generation`,
      );
    }
    const leaseOwner =
      raw.lease_owner === null || raw.lease_owner === undefined
        ? null
        : agentId(
            raw.lease_owner,
            `shared_goal_alignment.claims[${index}].lease_owner`,
          );
    // Lease facts travel as a pair: an active lease must carry both a
    // positive generation and a valid owner. A half-present pair is corrupt
    // authority — accepting it would let the reducer project the broken
    // lease as conflict-free, so decode rejects it at the boundary.
    if ((leaseEpoch !== null) !== (leaseOwner !== null)) {
      throw new EffectRuntimeRequestError(
        `shared_goal_alignment.claims[${index}].lease_epoch and lease_owner must be both present for an active lease or both null without one`,
      );
    }
    return {
      todo_id: todoId(raw.todo_id, `shared_goal_alignment.claims[${index}].todo_id`),
      claimed_by: claimedBy,
      lease_epoch: leaseEpoch,
      lease_owner: leaseOwner,
    };
  });
  requireNoDuplicates(
    claims.map((claim) => claim.todo_id),
    "shared_goal_alignment.claims",
  );
  return claims;
}

function decodeUnclaimedEligible(
  value: unknown,
  claimedTodoIds: Set<string>,
): UnclaimedEligibleFacts[] {
  if (!Array.isArray(value)) {
    throw new EffectRuntimeRequestError(
      "shared_goal_alignment.unclaimed_eligible must be an array",
    );
  }
  const items: UnclaimedEligibleFacts[] = value.map((item, index) => {
    const raw = requireJsonObject(
      item,
      `shared_goal_alignment.unclaimed_eligible[${index}]`,
    );
    const todoIdValue = todoId(
      raw.todo_id,
      `shared_goal_alignment.unclaimed_eligible[${index}].todo_id`,
    );
    if (claimedTodoIds.has(todoIdValue)) {
      throw new EffectRuntimeRequestError(
        `shared_goal_alignment.unclaimed_eligible[${index}].todo_id is already claimed by the projected agent`,
      );
    }
    const taskClass = requireNonEmptyString(
      raw.task_class,
      `shared_goal_alignment.unclaimed_eligible[${index}].task_class`,
    )
      .trim()
      .toLowerCase();
    if (taskClass !== "advancement_task") {
      throw new EffectRuntimeRequestError(
        `shared_goal_alignment.unclaimed_eligible[${index}].task_class must be advancement_task`,
      );
    }
    const actionKind = optionalNonEmptyString(
      raw.action_kind,
      `shared_goal_alignment.unclaimed_eligible[${index}].action_kind`,
    );
    const decoded: UnclaimedEligibleFacts = {
      todo_id: todoIdValue,
      task_class: "advancement_task",
    };
    if (actionKind !== null) decoded.action_kind = actionKind;
    return decoded;
  });
  requireNoDuplicates(
    items.map((item) => item.todo_id),
    "shared_goal_alignment.unclaimed_eligible",
  );
  return items;
}

function decodePeerClaimedBoundTodoIds(
  value: unknown,
  claimedTodoIds: Set<string>,
): string[] {
  const todoIds = requireStringArray(
    value,
    "shared_goal_alignment.peer_claimed_bound_todo_ids",
  ).map((item, index) =>
    todoId(item, `shared_goal_alignment.peer_claimed_bound_todo_ids[${index}]`),
  );
  for (const todoIdValue of todoIds) {
    if (claimedTodoIds.has(todoIdValue)) {
      throw new EffectRuntimeRequestError(
        "shared_goal_alignment.peer_claimed_bound_todo_ids cannot include a Todo already claimed by the projected agent",
      );
    }
  }
  requireNoDuplicates(
    todoIds,
    "shared_goal_alignment.peer_claimed_bound_todo_ids",
  );
  return todoIds;
}

function driftFacts(
  sourceBasis: SourceBasisFacts,
  frontier: FrontierBasisFacts,
): SharedGoalAlignmentDriftFact[] {
  const facts: SharedGoalAlignmentDriftFact[] = [];
  const comparable =
    sourceBasis.revision_basis === "state_event_log" &&
    frontier.basis_source === "state_event_log" &&
    frontier.based_on_state_event_sequence !== null;
  if (
    comparable &&
    frontier.based_on_state_event_sequence! < sourceBasis.state_event_basis_sequence
  ) {
    facts.push("frontier_basis_behind");
  }
  return facts;
}

function conflictFacts(
  sourceBasis: SourceBasisFacts,
  frontier: FrontierBasisFacts,
  claims: AlignmentClaimFacts[],
  peerClaimedBoundTodoIds: string[],
  openLaneReplanObligationRequired: boolean,
): SharedGoalAlignmentConflictFact[] {
  const facts: SharedGoalAlignmentConflictFact[] = [];
  if (
    frontier.basis_source === "unbound" ||
    sourceBasis.revision_basis === "markdown_active_state"
  ) {
    facts.push("frontier_basis_unverifiable");
  }
  if (
    claims.some(
      (claim) => claim.lease_owner !== null && claim.lease_owner !== claim.claimed_by,
    )
  ) {
    facts.push("lease_owner_mismatch");
  }
  if (openLaneReplanObligationRequired) {
    facts.push("open_lane_replan_obligation");
  }
  if (peerClaimedBoundTodoIds.length > 0) {
    facts.push("peer_claimed_lane_conflict");
  }
  return facts;
}

export function decodeSharedGoalAlignmentRequest(
  value: unknown,
): SharedGoalAlignmentRequest {
  let request = requireJsonObject(value, "shared_goal_alignment request");
  if (
    request.schema_version !== SHARED_GOAL_ALIGNMENT_REQUEST_SCHEMA_VERSION
  ) {
    throw new EffectRuntimeRequestError(
      "shared goal alignment request schema mismatch",
    );
  }
  const goalIdValue = requireNonEmptyString(
    request.goal_id,
    "shared_goal_alignment.goal_id",
  );
  if (!GOAL_ID_PATTERN.test(goalIdValue)) {
    throw new EffectRuntimeRequestError(
      "shared_goal_alignment.goal_id must be a safe single-segment goal id (no \"/\", \"\\\", whitespace, or path traversal)",
    );
  }
  const agentIdValue = agentId(request.agent_id, "shared_goal_alignment.agent_id");
  if (request.work_items !== undefined) {
    for (const key of ["claims", "frontier_counts", "unclaimed_eligible", "peer_claimed_bound_todo_ids"]) {
      if (request[key] !== undefined) throw new EffectRuntimeRequestError("work_items cannot mix with preselected alignment facts");
    }
    request = {...request, ...sharedGoalWorkFacts(request.work_items, agentIdValue, request.observed_at)};
  }
  const sourceBasis = decodeSourceBasis(request.source_basis);
  const frontier = decodeFrontierBasis(request.frontier_basis, sourceBasis);
  const claims = decodeClaims(request.claims, agentIdValue);
  const claimedTodoIds = new Set(claims.map((claim) => claim.todo_id));
  const openLaneReplanObligationRequired = requireBoolean(
    request.open_lane_replan_obligation_required,
    "shared_goal_alignment.open_lane_replan_obligation_required",
  );
  return {
    schema_version: SHARED_GOAL_ALIGNMENT_REQUEST_SCHEMA_VERSION,
    goal_id: goalIdValue,
    agent_id: agentIdValue,
    source_basis: sourceBasis,
    frontier_basis: frontier,
    frontier_counts: decodeFrontierCounts(request.frontier_counts),
    claims,
    unclaimed_eligible: decodeUnclaimedEligible(
      request.unclaimed_eligible,
      claimedTodoIds,
    ),
    peer_claimed_bound_todo_ids: decodePeerClaimedBoundTodoIds(
      request.peer_claimed_bound_todo_ids,
      claimedTodoIds,
    ),
    open_lane_replan_obligation_required: openLaneReplanObligationRequired,
  };
}

/** Project one read-only per-Agent alignment view from typed facts. */
export function projectSharedGoalAlignment(value: unknown): SharedGoalAlignment {
  const request = decodeSharedGoalAlignmentRequest(value);
  return {
    schema_version: SHARED_GOAL_ALIGNMENT_SCHEMA_VERSION,
    goal_id: request.goal_id,
    agent_id: request.agent_id,
    source_basis: request.source_basis,
    frontier_basis: request.frontier_basis,
    frontier_counts: request.frontier_counts,
    unclaimed_eligible_work: request.unclaimed_eligible.map((item) => ({
      todo_id: item.todo_id,
      claim_required_before_work: true,
    })),
    drift_facts: driftFacts(request.source_basis, request.frontier_basis),
    conflict_facts: conflictFacts(
      request.source_basis,
      request.frontier_basis,
      request.claims,
      request.peer_claimed_bound_todo_ids,
      request.open_lane_replan_obligation_required,
    ),
    read_only: true,
  };
}
